crypto/crypto_test.go

```diff
@@ -1,88 +1,60 @@
-// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
-
-// THIS WAS TAKEN FROM github.com/ethereum/go-ethereum MODIFY CAREFULLY
-
-package crypto
+package crypto_test
 
 import (
-	"crypto/ecdsa"
 	"testing"
 
-	"gx/ipfs/QmZp3eKdYQHHAneECmeK6HhiMwTPufmjC8DuuaGKv3unvx/blake2b-simd"
-
 	"gx/ipfs/QmPVkJMTeRC6iBByPWdrRkD3BE5UXsj5HPzb4kPqL186mS/testify/assert"
-	"gx/ipfs/QmPVkJMTeRC6iBByPWdrRkD3BE5UXsj5HPzb4kPqL186mS/testify/require"
-)
 
-func TestToECDSAErrors(t *testing.T) {
-	assert := assert.New(t)
-
-	_, err := HexToECDSA("0000000000000000000000000000000000000000000000000000000000000000")
-	assert.Error(err)
-
-	_, err = HexToECDSA("ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff")
-	assert.Error(err)
-}
+	"github.com/filecoin-project/go-filecoin/crypto"
+)
 
-func TestSign(t *testing.T) {
+func TestGenerateKey(t *testing.T) {
 	assert := assert.New(t)
-	require := require.New(t)
 
-	// Generate a key pair and convert the public key to bytes for easy comparison
-	key, err := GenerateKey()
+	sk, err := crypto.GenerateKey()
 	assert.NoError(err)
-	ecdasKey, ok := key.Public().(*ecdsa.PublicKey)
-	require.True(ok)
-	// This is the key we will compare recovered keys against
-	expPub := ECDSAPubToBytes(ecdasKey)
 
-	// Sign a message using the key we generated previously
-	msg := blake2b.Sum256([]byte("have you considered not asking passive aggressive questions?"))
-	sig, err := Sign(msg[:], key)
-	assert.NoError(err)
-
-	// recover a public key from a message and a signature, it should be
-	// equal to the previously generated public key
-	recPub, err := Ecrecover(msg[:], sig)
-	assert.NoError(err)
-	assert.Equal(expPub, recPub)
+	assert.Equal(len(sk), 32)
+	assert.NotEqual(sk[0], 0)
 
-	// should be equal to SigToPub since Ecrecover calls SigToPub under the hood
-	recoveredPub2, err := SigToPub(msg[:], sig)
-	assert.NoError(err)
-	recPub2 := ECDSAPubToBytes(recoveredPub2)
-	assert.Equal(expPub, recPub2)
+	msg := make([]byte, 32)
+	msg[0] = 1
 
-	// Basic sanity check, don't sign the message, pub keys should be different
-	evilMsg := blake2b.Sum256([]byte("being the detective in a crime movie where you are also the murderer..debugging"))
-	recPub, err = Ecrecover(evilMsg[:], sig)
+	digest, err := crypto.Sign(sk, msg)
 	assert.NoError(err)
-	assert.NotEqual(expPub, recPub)
-
-	recoveredPub2, err = SigToPub(evilMsg[:], sig)
+	assert.Equal(len(digest), 65)
+	pk := crypto.PublicKey(sk)
+
+	// valid signature
+	assert.True(crypto.Verify(pk, msg, digest))
+
+	// invalid signature - different message (too short)
+	assert.False(crypto.Verify(pk, msg[3:], digest))
+
+	// invalid signature - different message
+	msg2 := make([]byte, 32)
+	copy(msg2, msg)
+	msg2[0] = 1
+	msg2[1] = 255
+	assert.False(crypto.Verify(pk, msg2, digest))
+
+	// invalid signature - different digest
+	digest2 := make([]byte, 65)
+	copy(digest2, digest)
+	digest2[0] = 1
```

Comment:
so does the test fail if `digest[0:2] == []byte{1,255}` ? should happen every 1/2^16 runs

Reply:
sure does:
```
frrist@pine ~/s/g/f/g/crypto (switch-crypto)> go test -count=25600 ./...
--- FAIL: TestGenerateKey (0.00s)
    crypto_test.go:46: 
                Error Trace:    crypto_test.go:46
                Error:          Should be false
                Test:           TestGenerateKey
FAIL
FAIL    github.com/filecoin-project/go-filecoin/crypto  30.080s
```
will fix

Reply:
ups